server/src-lib/Hasura/Server/Auth/WebHook.hs

```diff
@@ -0,0 +1,150 @@
+import           Hasura.Server.Version     (HasVersion)
+
+import qualified Data.Aeson                as J
+import qualified Data.ByteString.Lazy      as BL
+import qualified Data.HashMap.Strict       as Map
+import qualified Data.Text                 as T
+import qualified Network.HTTP.Client       as H
+import qualified Network.HTTP.Types        as N
+import qualified Network.Wreq              as Wreq
+
+import           Data.Parser.CacheControl
+import           Data.Parser.Expires
+import           Hasura.HTTP
+import           Hasura.Logging
+import           Hasura.Prelude
+import           Hasura.RQL.Types
+import           Hasura.Server.Logging
+import           Hasura.Server.Utils
+
+
+data AuthHookType
+  = AHTGet
+  | AHTPost
+  deriving (Eq)
+
+instance Show AuthHookType where
+  show AHTGet  = "GET"
+  show AHTPost = "POST"
+
+isPost :: AuthHookType -> Bool
+isPost AHTGet  = False
+isPost AHTPost = True
+
+
+data AuthHookG a b
+  = AuthHookG
+  { ahUrl  :: !a
+  , ahType :: !b
+  } deriving (Show, Eq)
+
+type AuthHook = AuthHookG T.Text AuthHookType
+
+withAuthHook :: AuthHook -> (T.Text -> a) -> (T.Text -> a) -> a
+withAuthHook (AuthHookG u t) ifGet ifPost =
+  bool (ifGet u) (ifPost u) $ isPost t
+
+hookMethod :: AuthHook -> N.StdMethod
+hookMethod = bool N.GET N.POST . isPost . ahType
+
+
+-- | Extracts 'UserInfo' from the webhook answer.
+--   This uses the provided manager to make the call, extracts
+--   information such as response body and status, and delegates to a
+--   local 'mkUserInfoFromResp'.
+userInfoFromAuthHook
+  :: (HasVersion, MonadIO m, MonadError QErr m)
+  => Logger Hasura
+  -> H.Manager
+  -> AuthHook
+  -> [N.Header]
+  -> m (UserInfo, Maybe UTCTime)
+userInfoFromAuthHook logger manager hook reqHeaders = do
+  res  <- liftIO $ try $ withAuthHook hook withGET withPOST
+  resp <- either logAndThrow return res
+  let status   = resp ^. Wreq.responseStatus
+      respBody = resp ^. Wreq.responseBody
+  mkUserInfoFromResp logger (ahUrl hook) (hookMethod hook) status respBody
+
+  where
+    withGET  url = Wreq.getWith  (mkOptions filteredHeaders) $ T.unpack url
+    withPOST url = Wreq.postWith (mkOptions [contentType]) (T.unpack url) $
+      object ["headers" J..= postHdrsPayload]
+    mkOptions = wreqOptions manager
+    contentType = ("Content-Type", "application/json")
+    postHdrsPayload = J.toJSON $ Map.fromList $ hdrsToText reqHeaders
+
+    logAndThrow err = do
+      unLogger logger $
+        WebHookLog LevelError Nothing (ahUrl hook) (hookMethod hook)
+        (Just $ HttpException err) Nothing Nothing
+      throw500 "Internal Server Error"
+
+    filteredHeaders = flip filter reqHeaders $ \(n, _) ->
+      n `notElem` commonClientHeadersIgnored
+
+
+-- | Processes the result of the wehook call, and tries to construct a
+--   'UserInfo'. The expiration time is retrieved from either the
+--   CacheControl or Expires header.
```

Comment:
I’d delete this comment, personally—I think it mostly just describes what the function does, and I think the name `mkUserInfoFromResponse` captures most of that.

Reply:
Done!